nucypher/crypto/powers.py

```diff
@@ -67,69 +76,115 @@ def consume_power_up(self, power_up):
             raise TypeError(
                 ("power_up must be a subclass of CryptoPowerUp or an instance "
                  "of a CryptoPowerUp subclass."))
-        self._power_ups[power_up_class] = power_up_instance
+        self.__power_ups[power_up_class] = power_up_instance
 
         if power_up.confers_public_key:
             self.public_keys[power_up_class] = power_up_instance.public_key()
 
     def power_ups(self, power_up_class):
         try:
-            return self._power_ups[power_up_class]
+            return self.__power_ups[power_up_class]
         except KeyError:
             raise power_up_class.not_found_error
 
 
-class CryptoPowerUp(object):
+class CryptoPowerUp:
     """
     Gives you MORE CryptoPower!
     """
     confers_public_key = False
 
+    def activate(self, *args, **kwargs):
+        return
+
 
-class BlockchainPower(CryptoPowerUp):
+class TransactingPower(CryptoPowerUp):
     """
     Allows for transacting on a Blockchain via web3 backend.
     """
-    not_found_error = NoBlockchainPower
+    not_found_error = NoTransactingPower
 
-    def __init__(self, blockchain: 'Blockchain', account: str, device = None) -> None:
+    class NoBlockchainConnection(PowerUpError):
+        pass
+
+    class AccountLocked(PowerUpError):
+        pass
+
+    class InvalidSigningRequest(PowerUpError):
+        pass
+
+    def __init__(self,
+                 blockchain,
+                 account: str,
+                 password: str = None):
         """
-        Instantiates a BlockchainPower for the given account id.
+        Instantiates a TransactingPower for the given checksum_address.
         """
         self.blockchain = blockchain
+        if blockchain.is_connected:
+            self.client = blockchain.client
+        else:
+            self.client = NO_BLOCKCHAIN_CONNECTION
+
         self.account = account
-        self.device = device
-        self.is_unlocked = False
+        self.device = True if not password else False
+        self.__password = password
+        self.__unlocked = False
+
+    def __del__(self):
+        self.lock_account()
+
+    @property
+    def is_unlocked(self) -> bool:
+        return self.__unlocked
+
+    @property
+    def is_active(self) -> bool:
+        """Returns True if the blockchain currently has this transacting power attached."""
+        return self.blockchain.transacting_power == self
+
+    def activate(self, password: str = None):
+        """Be Consumed"""
+        self.blockchain.connect()
+        self.client = self.blockchain.client       # Connect
+        self.unlock_account(password=password)     # Unlock
```

Comment:
There's a pattern here that I don't understand and that is causing a bug, which I suffered yesterday it when trying to deploy on a real geth node (i.e., not `--dev`) with non-insecure accounts. 

The `TransactingPower` is first inited with a password, but also there's a new password here in `activate`. That's somewhat confusing: was the intention to use the first one or the second one? More problematic is that the second password seems currently to override the first one (?), and on top of that, the second password is optional (??). So the recurrent pattern of doing:

```python
foo.transacting_power = TransactingPower(blockchain=some_blockchain,
                                         password=some_password,
                                         account=some_account)
foo.transacting_power.activate()
```
...will simply try to unlock the account with no password, regardless of being inited with `some_password`. Incidentally, this works with unprotected accounts, which are the default accounts in ganache, `geth --dev`, and EthereumTester, AFAIK. I currently find 36 matches of `transacting_power.activate()` (i.e., `password=None`) through the codebase, which of course wasn't giving any trouble since we only use ganache, `geth --dev`, and EthereumTester so far. BTW, this problem is analog to what we found on #1105. 

I have a dumb fix in fbe11384ade00ee9acafd97975236b7d7592cddf, although you perhaps prefer a clever solution.

Reply:
:heavy_check_mark:  - Good Catch.  We need further integration testing with client keyings to prevent future regression.